osbenchmark/aggregator.py

```diff
@@ -0,0 +1,216 @@
+        self.config.add(config.Scope.applicationOverride, "workload", "latency.percentiles", first_test_execution.latency_percentiles)
+        self.config.add(config.Scope.applicationOverride, "workload", "throughput.percentiles", first_test_execution.throughput_percentiles)
+
+        loaded_workload = workload.load_workload(self.config)
+        test_procedure = loaded_workload.find_test_procedure_or_default(first_test_execution.test_procedure)
+
+        test_execution = metrics.create_test_execution(self.config, loaded_workload, test_procedure, first_test_execution.workload_revision)
+        test_execution.add_results(aggregated_results)
+        test_execution.distribution_version = test_exe.distribution_version
+        test_execution.revision = test_exe.revision
+        test_execution.distribution_flavor = test_exe.distribution_flavor
+        test_execution.provision_config_revision = test_exe.provision_config_revision
+
+        return test_execution
+
+    # calculate weighted averages for task metrics
+    def calculate_weighted_average(self, task_metrics: Dict[str, List[Any]], iterations: int) -> Dict[str, Any]:
+        weighted_metrics = {}
+
+        for metric, values in task_metrics.items():
+            weighted_metrics[metric] = {}
+            if isinstance(values[0], dict):
+                for item_key in values[0].keys():
+                    if item_key == 'unit':
+                        weighted_metrics[metric][item_key] = values[0][item_key]
+                    else:
+                        item_values = [value.get(item_key, 0) for value in values]
+                        if iterations > 1:
+                            weighted_sum = sum(value * iterations for value in item_values)
+                            total_iterations = iterations * len(values)
+                            weighted_metrics[metric][item_key] = weighted_sum / total_iterations
+                        else:
+                            weighted_metrics[metric][item_key] = sum(item_values) / len(item_values)
+            else:
+                if iterations > 1:
+                    weighted_sum = sum(value * iterations for value in values)
+                    total_iterations = iterations * len(values)
+                    weighted_metrics[metric] = weighted_sum / total_iterations
+                else:
+                    weighted_metrics[metric] = sum(values) / len(values)
+        return weighted_metrics
+
+    # verify that all test executions have the same workload
+    def compatibility_check(self, test_store) -> None:
+        first_test_execution = test_store.find_by_test_execution_id(list(self.test_executions.keys())[0])
+        workload = first_test_execution.workload
+        for id in self.test_executions.keys():
+            test_execution = test_store.find_by_test_execution_id(id)
+            if test_execution:
+                if test_execution.workload != workload:
```

Comment:
Aside from workload, we should also consider checking if the first test execution id's test_procedure matches the rest. Reason being some workloads have multiple test_procedures, such as NYC Taxis: https://github.com/opensearch-project/opensearch-benchmark-workloads/blob/main/nyc_taxis/test_procedures/default.json

Reply:
Added a check for this👍